google/cloud/pubsublite/publisher_connection.cc

```diff
@@ -0,0 +1,208 @@
+namespace pubsublite {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+using google::cloud::version_major;
+using google::cloud::version_minor;
+
+using google::cloud::internal::Base64Encoder;
+using google::cloud::internal::MakeBackgroundThreadsFactory;
+
+using google::cloud::pubsub::PublisherConnection;
+using google::cloud::pubsub_internal::ContainingPublisherConnection;
+
+using google::cloud::pubsublite_internal::AlarmRegistryImpl;
+using google::cloud::pubsublite_internal::AsyncSleeper;
+using google::cloud::pubsublite_internal::BatchingOptions;
+using google::cloud::pubsublite_internal::ClientMetadata;
+using google::cloud::pubsublite_internal::CreateDefaultAdminServiceStub;
+using google::cloud::pubsublite_internal::CreateDefaultPublisherServiceStub;
+using google::cloud::pubsublite_internal::DefaultPublishMessageTransformer;
+using google::cloud::pubsublite_internal::DefaultRoutingPolicy;
+using google::cloud::pubsublite_internal::MakeLocation;
+using google::cloud::pubsublite_internal::MakeStreamFactory;
+using google::cloud::pubsublite_internal::MultipartitionPublisher;
+using google::cloud::pubsublite_internal::PartitionPublisher;
+using google::cloud::pubsublite_internal::Publisher;
+using google::cloud::pubsublite_internal::PublisherConnectionImpl;
+using google::cloud::pubsublite_internal::
+    ResumableAsyncStreamingReadWriteRpcImpl;
+using google::cloud::pubsublite_internal::RetryPolicyFactory;
+using google::cloud::pubsublite_internal::StreamInitializer;
+using google::cloud::pubsublite_internal::StreamRetryPolicy;
+
+using google::cloud::pubsublite::v1::InitialPublishRequest;
+using google::cloud::pubsublite::v1::PublishRequest;
+using google::cloud::pubsublite::v1::PublishResponse;
+
+using google::protobuf::Struct;
+using google::protobuf::Value;
+
+BatchingOptions CreateBatchingOptions(Options const& opts) {
+  BatchingOptions batching_options;
+  if (opts.has<MaxBatchMessagesOption>()) {
+    batching_options.set_maximum_batch_message_count(
+        opts.get<MaxBatchMessagesOption>());
+  }
+  if (opts.has<MaxBatchBytesOption>()) {
+    batching_options.set_maximum_batch_bytes(opts.get<MaxBatchBytesOption>());
+  }
+  if (opts.has<PublishFlushAlarmPeriodOption>()) {
+    batching_options.set_alarm_period(
+        opts.get<PublishFlushAlarmPeriodOption>());
+  }
+  return batching_options;
+}
+
+StatusOr<std::string> GetEndpoint(std::string const& location) {
+  auto parsed_loc = MakeLocation(location);
+  if (!parsed_loc) return parsed_loc.status();
+  return absl::StrCat(parsed_loc->GetCloudRegion().ToString(),
+                      "-pubsublite.googleapis.com");
+}
+
+std::string GetSerializedContext(std::string const& framework) {
+  Struct context;
+  auto& metadata_map = *context.mutable_fields();
+  Value lang;
+  lang.set_string_value("CPP");
+  metadata_map["language"] = std::move(lang);
+  Value framework_val;
+  framework_val.set_string_value(framework);
+  metadata_map["framework"] = framework_val;
```

Comment:
anywhere you do a map accessor: it creates an implicit instance. just use for example:

metadata_map["framework"].set_string_value(framework);